packages/ember-simple-auth/addon/services/session.js

```diff
@@ -212,4 +283,99 @@ export default Service.extend(Evented, {
 
     return session.invalidate(...arguments);
   },
+
+  /**
+    Checks whether the session is authenticated and if it is not, transitions
+    to the specified route or invokes the specified callback.
+
+    If a transition is in progress and is aborted, this method will save it in the
+    session service's
+    {{#crossLink "SessionService/attemptedTransition:property"}}{{/crossLink}}
+    property so that  it can be retried after the session is authenticated. If
+    the transition is aborted in Fastboot mode, the transition's target URL
+    will be saved in a `ember_simple_auth-redirectTarget` cookie for use by the
+    browser after authentication is complete.
+
+    @method requireAuthentication
+    @param {Transition} transition A transition that triggered the authentication requirement or null if the requirement originated independently of a transition
+    @param {String|Function} routeOrCallback The route to transition to in case that the session is not authenticated or a callback function to invoke in that case
+    @return {Boolean} true when the session is authenticated, false otherwise
```

Comment:
So this method and `prohibitAuthentication` are more or less directly replacements for the `AuthenticatedRouteMixin` and `UnauthenticatedRouteMixin`. The logic in these mixins is completely sync as well since all they do is check whether the session is authenticated or not and if that's the case or not, redirect appropriately. I think what you're suggesting is turning the session's `isAuthenticated` property into something that's async but I'd very much like to avoid that. If you cannot synchronously know whether the session is authenticated or not that's a huge change to ESA's API and probably introduces significant complexity.

If you want to check whether the authentication data you have stored in the session is still valid before every visit to a protected route, you could of course not use the `requireAuthentication` method but roll your own that makes a request that validates the token every time. I think the approach that we've been taking in ESA for years though (if we have auth data in the session, assume it's valid until we learn it is not, e.g. by receiving a 401 response) has proven to work pretty well.

Reply:
> I think what you're suggesting is turning the session's `isAuthenticated` property into something that's async

You got me wrong here. `isAuthenticated` should be still a sync property. It should still reflect if valid authentication information are available. I only want to add official support for transparently establishing a local session if a SSO session exists.

Ember Simple Auth currently assumes that there are only two states:

1. The app has valid authentication credentials locally, which means that the user is authenticated.
2. The app does not have valid authentication credentials locally, which means the user is not authenticated.

A single sign-on service decouples the two states: The app may not have valid authentication credentials (e.g. an access token) locally but the user may have a SSO session. In that scenario the local session should be established transparently to the user.

The `requireAuthentication` method would be a good place to do this. An implementation might look like this:

```js
async requireAuthentication() {
  let { isAuthenticated } = this;

  // check if local session exists
  if (!isAuthenticated) {
    try {
      // try to establish local session transparently to the user
      await this.authenticate();
    } catch (error) {
      // wasn't able to establish a local session without user interaction
      // use default implementation to redirect user to SSO login page
      await super.requireAuthentication(...arguments);

      return false;
    }
  }

  return true;
}
```

Please note that is mainly a change to the documentation. If I didn't missed something it's only about two changes to the docs: 

1. Change the return values for `requireAuthentication` and `prohibitAuthentication` to `Promise`.
2. Change the integration examples in guides like the following:
   ```patch
   // app/routes/index.js
   import Route from '@ember/routing';
   import { inject as service } from '@ember/service';
   
   export default IndexRoute extends Route {
   -  beforeModel(transition) {
   +  async beforeModel(transition) {
   -    this.session.requireAuthentication(transition, 'login');
   +    await this.session.requireAuthentication(transition, 'login');
     },
   }
   ```

Reply:
I think you're describing a very specific case here and I don't think that should be covered by ESA by default as it would not actually be relevant to > 90% of its users. Also there seems to be a mis-conception here: what you're describing really is trying to **authenticate** the session before calling `requireAuthentication` which is totally fine and you can do that but that's what ESA has authenticators for. Also an authenticator isn't required to store any credentials in the session (which you'd not even be able to do in case of a http-only cookie-based session). So I think what you're describing is really just an SSO **authenticator**. You could do something like this (without `requireAuthentication` being async):

```js
async beforeModel(transition) {
  await this.tryAuthenticationViaSSO();
  this.session.requireAuthentication(transition, 'login');
}

tryAuthenticationViaSSO() {
  if (!this.session.isAuthenticated) {
    try {
      await this.session.authenticate('authenticator:sso')
    } catch (e) {
      // redirect to login here or let requireAuthentication above handle that
    }
  }
}
```

The current implementation of `requireAuthentication` works essentially the same way the `AuthenticatedRouteMixin` works and I think we don't need to change that to accomodate the case you describe. The general strategy in ESA is to provide the primitives one needs to implement all kinds of (or really any) auth patterns and flows plus providing sensible default implementations for typical scenarios.

Reply:
Resolving for now – @jelhan feel free to unresolve if you think this remains relevant.